instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py

```diff
@@ -136,74 +189,57 @@ def get_from_grpc_metadata(metadata, key) -> List[str]:
         else:
             yield
 
-    def _start_span(self, method):
-        span = self._tracer.start_as_current_span(
-            name=method, kind=trace.SpanKind.SERVER
+    def _start_span(self, handler_call_details, context):
+
+        attributes = {
+            "grpc.method": handler_call_details.method,
+        }
+
+        metadata = dict(context.invocation_metadata())
+        if "user-agent" in metadata:
+            attributes["grpc.user_agent"] = metadata["user-agent"]
+
+        # Split up the peer to keep with how other telemetry sources
+        # do it.  This looks like ipv6:[::1]:57284 or ipv4:127.0.0.1:57284.
+        host, port = context.peer().split(":", 1)[1].rsplit(":", 1)
+
+        # other telemetry sources convert this, so we will too
+        if host in ("[::1]", "127.0.0.1"):
+            host = "localhost"
+
+        attributes.update({"net.peer.name": host, "net.peer.port": port})
+
+        return self._tracer.start_as_current_span(
+            name=handler_call_details.method,
+            kind=trace.SpanKind.SERVER,
+            attributes=attributes,
+        )
+
+    def intercept_service(self, continuation, handler_call_details):
+        def telemetry_wrapper(behavior, request_streaming, response_streaming):
+            def telemetry_interceptor(request_or_iterator, context):
+
+                with self._set_remote_context(context):
+                    with self._start_span(
+                        handler_call_details, context
+                    ) as span:
+                        # wrap the context
+                        context = _OpenTelemetryServicerContext(context, span)
+
+                        # And now we run the actual RPC.
+                        try:
+                            return behavior(request_or_iterator, context)
+                        except Exception as error:
+                            # Bare exceptions are likely to be gRPC aborts, which
+                            # we handle in our context wrapper.
+                            # Here, we're interested in uncaught exceptions.
+                            # pylint:disable=unidiomatic-typecheck
+                            if type(error) != Exception:
```

Comment:
Concerned that this might result in some non-gRPC abort exceptions being ignored. Do gRPC aborts raise `Exception` instead of some specialized type like `GRPCAbortException`? Is there any other way to test for this like exception message?

Also, what if we did record gRPC abort exceptions? Is that unexpected? Does gRPC use exceptions as flow control instead of error in this case?

Reply:
Yeah unfortunately once we're at this point, if the code called `grpc.abort()` the exception here is a bare `Exception` with no useful details at all (see here: https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_server.py#L319), and there's no other place to capture it.  I've tried a whole bunch of ways to get this over the past year or so, and this is the best I've come up with.

If `abort()` was not called, the exception here is whatever was raised, and ultimately it'll get wrapped up in a `grpc.RpcError` with status = `UNKNOWN` (HTTP 500 equivalent) and the client will receive that result.

I've certainly tried to find some better way to handle this, but there's just not any other information available to the interceptor here.

Reply:
From the code you linked, it seems the function is setting always sets `code` to a non-OK value indicating this is always an error condition. In that case, may be we should always record it with `span.record_exception()`?

Reply:
Okay that works for me - I'm not as familiar with Otel itself as you guys :)

Reply:
Oh actually - I forgot how this actually works here.  A bare exception is going to usually be a grpc abort, but not all of those are actually exceptions, there are user errors such as the equivalents of HTTP 400 series which aren't really server exceptions.  Do we want to mark those as exceptions in Otel?  I wouldn't think so...

I handle grpc aborts in the `abort()` call in `_OpenTelemetryServicerContext` (and can certainly use `span.record_exception()` for the non-user stuff there), this block here was intended to catch anything that is NOT an abort, and therefore a real exception, which are unlikely to be bare `Exception` (because of good practice).

There's no other way to distinguish a true bare `Exception` from a grpc abort here.

Reply:
If it always represents an error condition, we should add it as an exception. Only case we should not add it would be if exception is being used here as flow control to jump out of a function to it's grand-parent.

Reply:
A HTTP 404 is treated as an exception in Otel?

Reply:
404s are not treated as errors generally, no. Does this case happen in case of a 404?

Reply:
Yes - gRPC unfortunately uses exceptions as program flow, so all 400-like errors which you cause by running `grpc.abort()` show up here, as bare exceptions.  That's the part I was trying to work around and why I didn't treat those as exceptions.

I can add some checks for the "not user error" types of exceptions in the `abort()` call in `_OpenTelemetryServicerContext`, it's a small number of them that are actual exceptional conditions.

Reply:
OK, let's do it if it is easy enough and doesn't complicate the instrumentation too much. Otherwise I think it's fine the way you have it as it is not that likely that users will raise bare Exception (but still possible)

Reply:
Argh this is complicated - at the point where I'd like to record this as an exception, there's not a useful exception to record, because aborts aren't actually exceptions.

Following conventions in the docs on HTTP details in traces, I'll set the `Status` properly.

Reply:
Okay did a bit more digging - it's been a while since I wrote this. The parts in the `Context` wrapper where I set the status code line up with how Otel status codes work - they match 1:1 with the `grpc.StatusCode` values and so there's no change needed here to handle these cases.